src/runtime/virtcontainers/clh.go

```diff
@@ -1261,37 +1399,71 @@ func (clh *cloudHypervisor) bootVM(ctx context.Context) error {
 		}
 		clh.Logger().WithField("body", string(bodyBuf)).Debug("VM config")
 	}
-	_, err := cl.CreateVM(ctx, clh.vmconfig)
-	if err != nil {
-		return openAPIClientError(err)
-	}
 
-	info, err := clh.vmInfo()
-	if err != nil {
-		return err
-	}
+	if !clh.config.BootFromTemplate {
+		_, err := cl.CreateVM(ctx, clh.vmconfig)
+		if err != nil {
+			return openAPIClientError(err)
+		}
+	
+		info, err := clh.vmInfo()
+		if err != nil {
+			return err
+		}
+	
+		clh.Logger().Debugf("VM state after create: %#v", info)
+	
+		if info.State != clhStateCreated {
+			return fmt.Errorf("VM state is not 'Created' after 'CreateVM'")
+		}
 
-	clh.Logger().Debugf("VM state after create: %#v", info)
+		clh.Logger().Debug("Booting VM")
+		_, err = cl.BootVM(ctx)
+		if err != nil {
+			return openAPIClientError(err)
+		}
 
-	if info.State != clhStateCreated {
-		return fmt.Errorf("VM state is not 'Created' after 'CreateVM'")
-	}
+		info, err = clh.vmInfo()
+		if err != nil {
+			return err
+		}
+	
+		clh.Logger().Debugf("VM state after boot: %#v", info)
+	
+		if info.State != clhStateRunning {
+			return fmt.Errorf("VM state is not 'Running' after 'BootVM'")
+		}
+	} else {
+		clh.Logger().Info("Restoring VM from template")
 
-	clh.Logger().Debug("Booting VM")
-	_, err = cl.BootVM(ctx)
-	if err != nil {
-		return openAPIClientError(err)
-	}
+		vmStoreDir := filepath.Join(clh.config.VMStorePath, clh.id)
+		symlink := vmStoreDir + "/state.json"
+		err := os.Symlink(clh.config.DevicesStatePath, symlink)
+		if err != nil {
+			return err
+		}
 
-	info, err = clh.vmInfo()
-	if err != nil {
-		return err
-	}
+		data, errJSON := json.MarshalIndent(clh.vmconfig, "", "\t")
+		if errJSON != nil {
+			return errJSON
+		}
+		vmConfigFile := vmStoreDir + "/config.json"
+		err = os.WriteFile(vmConfigFile, data, 0640)
+		if err != nil {
+			return err
+		}
 
-	clh.Logger().Debugf("VM state after boot: %#v", info)
+		err = clh.restoreFromTemplate(vmStoreDir)
+		if err != nil {
+			return openAPIClientError(err)
+		}
```

Comment:
From VMM/CLH perspective, its users (e.g. Kata) should only concern about the path of the snapshot (to store or restore). 

I am not following why the restoring path needs to reply on the internal information of the snapshot/restore functionality of Cloud Hypervisor, e.g. "config.json" and "state.json"? 

/cc @amshinde @fidencio maybe I am missing some context for kata's VM template.